M2/Macaulay2/m2/document.m2

```diff
@@ -515,38 +524,44 @@ getSignature Sequence := x -> (
+    outputs = apply(outputs, T -> if instance(T, Symbol) then value T else T);
+    (inputs, outputs))
 
 isOption := opt -> instance(opt, Option) and #opt == 2 and instance(opt#0, Symbol);
 
 processUsage := (tag, fn, o) -> (
-    if not o.?Usage and (o.?Inputs or o.?Outputs)
-    then error "document: Inputs or Outputs specified, but Usage not provided";
-    arg := if o.?Inputs then o.Inputs else {};
-    out := if o.?Outputs then o.Outputs else {};
-    (ino, inp) := toSequence values partition(isOption, arg, {true, false});
-    opt := getOptionDefaultValues tag.Key;
-    inoh:= new HashTable from ino;
-    if not isSubset(keys inoh, keys opt)
-    then error concatenate("not among the options for ", toString fn, ": ", between_", " keys (set keys inoh - set keys opt));
-    ino = join(ino, sortByName (keys opt - set keys inoh));
-    if o.?Usage then (
-	(inp', out') := getSignature tag.Key;
-	inp' = select(inp', T -> T =!= Nothing);
-	out' = select(out', T -> T =!= Nothing);
-	-- When T is not exported, its class evaluates to Symbol instead of Type
-	inp' = apply(inp', T -> if instance(T, Symbol) then value T else T);
-	out' = apply(out', T -> if instance(T, Symbol) then value T else T);
-	if out' === {Thing} then out' = {};		    -- not informative enough
-	if #inp === 0 then inp = inp';
-	if #out === 0 then out = out';
-	if #inp' =!= 0 then (
-	    if #inp =!= #inp' then error ("mismatched number of inputs in documentation for ", format tag);
-	    inp = apply(inp',inp,(T,v) -> T => v));
-	if #out' =!= 0 then (
-	    if #out =!= #out' then error ("mismatched number of outputs in documentation for ", format tag);
-	    out = apply(out',out,(T,v) -> T => v)));
-    apply((inp, out, ino), x -> apply(x, processSignature(tag, fn))))
+    if not o.?Usage then (
+	if o.?Inputs or o.?Outputs then error "document: Inputs or Outputs specified, but Usage not provided";
+	return ({}, {}, {}));
+    -- inputs, outputs, and options provided in the documentation
+    arguments := partition(isOption, if o.?Inputs then o.Inputs else {}, {false, true});
+    inputList  := arguments#false;
+    optionList := arguments#true;
+    outputList := if o.?Outputs then o.Outputs else {};
+    optionNames := keys hashTable optionList;
+    -- default inputs, outputs, and options of the function or method
+    (inputs0, outputs0) := getSignature tag.Key;
+    options0 := getOptionDefaultValues  tag.Key;
+    -- e.g. map(Module, Nothing, Matrix)
+    inputs0  = select( inputs0, T -> T =!= Nothing);
+    outputs0 = select(outputs0, T -> T =!= Nothing and T =!= Thing);
+    options0 = if options0 === true then optionNames else keys options0;
+    -- check that option names are valid
+    if not isSubset(optionNames, options0) then error("not among the options for ", format tag, ": ",
+	between_", " (optionNames - set options0));
+    optionList = join(optionList, sortByName (options0 - set optionNames));
```

Comment:
Should we sort all the options?  Right now, the documented options appear first in whatever order they were documented, and the undocumented appear afterwards, but sorted.

Reply:
I think the current behavior makes sense, because the author has control over the options important enough to document, and the rest (if there are any) are collected at the end rather than random empty gaps.

Reply:
Makes sense!